### ISSUE TITLE
disabled perfmark still cost CPU

### ISSUE THREAD

**Issue by ran-su** [perfmark/perfmark#132]
The `SecretPerfMarkImpl$PerfMarkImpl.getGen()` is called in every PerfMark methods and it supposed to be no cost after JIT. However, In our production environment, we are seeing some cost under this method, more specifically in the `SecretMethodHandleGenerator$MethodHandleGenerator.getGeneration()` method. The java6 and java7 packages are always included in our binaries, but the perfmark is rarely enable in the production environment.

The attached picture is a sample breakdown. The percent number in a block represents the CPU cost in that method itself and all methods it invokes.

![Screenshot 2021-06-07 3 31 14 PM](https://user-images.githubusercontent.com/1212561/121096359-ebd47500-c7a6-11eb-80f7-01050ef72650.png)

The environment are a mix of jdk8 and jdk11.

**Comment by carl-mastrangelo** [comment]
# TL;DR: This is normal, and likely a result of sampling the code before the JIT has finished generating optimized code.  The steady state will be zero overhead.

A likely cause for this is that these traces were taken at start up.   When the JVM starts, all code is executed in interpreter mode.   In this mode, the JVM is compiling code in the background, but it is not yet ready.  If the queue for the compiler thread is long, it may be stuck in interpreter mode for longer.   However, it may also be in Compiler mode 1 (a.k.a. C1), which is the step after the Interpreter.  Again, it may be stuck in this mode for a long time while it is waiting for the optimizing compiler  (C4, but confusingly also called C2) to kick in.   

I did some profiling of the MethodHandle Generator to see what it behaves like on different levels.  This was done using JDK 16, using the G1 GC.


## Optimized (Tier 4, C2)

```
Benchmark                                                                                        Mode  Cnt     Score     Error  Units
MethodHandleGeneratorBenchmarkTest.MethodHandleGeneratorBenchmark.getAndSetAndGetGeneration      avgt    5  1343.149 ± 205.585  ns/op
MethodHandleGeneratorBenchmarkTest.MethodHandleGeneratorBenchmark.getGeneration                  avgt    5     3.135 ±   0.046  ns/op
MethodHandleGeneratorBenchmarkTest.MethodHandleGeneratorBenchmark.ifEnabled                      avgt    5     0.768 ±   0.020  ns/op
MethodHandleGeneratorBenchmarkTest.MethodHandleGeneratorBenchmark.racyGetAndSetAndGetGeneration  avgt    5  2832.799 ± 171.987  ns/op
```

The relevant number is the `ifEnabled` benchmark.   The number is actually 0, as the `0.768 ns/op` is a benchmarking artifact.   Looking at the generated x86_64 assembly confirms that there is no code there. The benchmark method looks like:

```
  @Benchmark
  @BenchmarkMode(Mode.AverageTime)
  @OutputTimeUnit(TimeUnit.NANOSECONDS)
  public void ifEnabled() {
    if (isEnabled(getGeneration())) {
      Blackhole.consumeCPU(1000);
    }
  }
```

If we look at the benchmark with Async Profiler, which collects stack traces, we can see there is no code either:

```
          ns  percent  samples  top
  ----------  -------  -------  ---
  4969976326   99.60%      497  io.perfmark.java7.jmh_generated.MethodHandleGeneratorBenchmarkTest_MethodHandleGeneratorBenchmark_ifEnabled_jmhTest.ifEnabled_avgt_jmhStub
    20060662    0.40%        2  io.perfmark.testing.GeneratorBenchmark.getGeneration
```

The graph you posted shows that there are definitely stack frames that lead into the method handle generated code, so let's look again at interpreter mode.

## Interpreted

This mode would be used when the program is very early into booting up.  The results would be very slow, which would match what you are noticing in terms of CPU usage:

```
Benchmark                                                                                        Mode  Cnt      Score      Error  Units
MethodHandleGeneratorBenchmarkTest.MethodHandleGeneratorBenchmark.getAndSetAndGetGeneration      avgt    5  17980.293 ± 2543.317  ns/op
MethodHandleGeneratorBenchmarkTest.MethodHandleGeneratorBenchmark.getGeneration                  avgt    5   1236.723 ±   12.713  ns/op
MethodHandleGeneratorBenchmarkTest.MethodHandleGeneratorBenchmark.ifEnabled                      avgt    5   1265.270 ±    5.887  ns/op
MethodHandleGeneratorBenchmarkTest.MethodHandleGeneratorBenchmark.racyGetAndSetAndGetGeneration  avgt    5  22390.795 ± 2066.916  ns/op
```

However, looking more into what is going on during the intrepretter mode, we can see that the stack frames don't match the pprof graph you provided:


```
          ns  percent  samples  top
  ----------  -------  -------  ---
   850109192   17.04%       85  Interpreter
   440000408    8.82%       44  java.lang.invoke.MethodHandleImpl.isCompileConstant
   429972554    8.62%       43  java.lang.invoke.Invokers.checkCustomized
   359970131    7.21%       36  java.lang.invoke.LambdaForm$MH.0x0000000800c2e800.invoke
   269960076    5.41%       27  java.lang.invoke.MethodHandle.asType
   269916455    5.41%       27  io.perfmark.java7.SecretMethodHandleGenerator$MethodHandleGenerator.getGeneration
   250006352    5.01%       25  java.lang.invoke.LambdaForm$MH.0x0000000800c2c800.invoke
   239938005    4.81%       24  java.lang.invoke.MutableCallSite.getTarget
   230042528    4.61%       23  java.lang.invoke.Invokers.checkExactType
   210006371    4.21%       21  java.lang.invoke.LambdaForm$MH.0x0000000800c2e400.invoke_MT
   209968631    4.21%       21  java.lang.invoke.Invokers.checkGenericType
   189992178    3.81%       19  io.perfmark.testing.GeneratorBenchmark.ifEnabled
   179969021    3.61%       18  java.lang.invoke.MethodHandle.type
   170000181    3.41%       17  io.perfmark.testing.GeneratorBenchmark.getGeneration
   150084643    3.01%       15  java.lang.invoke.DirectMethodHandle$Holder.invokeVirtual
   140034940    2.81%       14  io.perfmark.testing.GeneratorBenchmark.isEnabled
   120035869    2.41%       12  java.lang.invoke.DirectMethodHandle.internalMemberName
    89993208    1.80%        9  _invokeBasic
    79995877    1.60%        8  java.lang.invoke.LambdaForm$MH.0x0000000800c2ec00.invoke
    59997753    1.20%        6  /home/carl/.gradle/jdks/jdk-16.0.1+9/lib/server/classes.jsa
    39983920    0.80%        4  io.perfmark.java7.jmh_generated.MethodHandleGeneratorBenchmarkTest_MethodHandleGeneratorBenchmark_ifEnabled_jmhTest.ifEnabled_avgt_jmhStub
    10007487    0.20%        1  clocksource_watchdog_[k]
```

We should see a lot of `isCompileConstant` in the results, but the provided graph doesn't match.   This implies to me that it is not getting stuck in the interpretter.

## Compiled (Tier 1, C1)

```
Benchmark                                                                                        Mode  Cnt     Score     Error  Units
MethodHandleGeneratorBenchmarkTest.MethodHandleGeneratorBenchmark.getAndSetAndGetGeneration      avgt    5   883.180 ±  63.953  ns/op
MethodHandleGeneratorBenchmarkTest.MethodHandleGeneratorBenchmark.getGeneration                  avgt    5    16.380 ±   2.451  ns/op
MethodHandleGeneratorBenchmarkTest.MethodHandleGeneratorBenchmark.ifEnabled                      avgt    5    16.413 ±   0.389  ns/op
MethodHandleGeneratorBenchmarkTest.MethodHandleGeneratorBenchmark.racyGetAndSetAndGetGeneration  avgt    5  1476.741 ± 572.736  ns/op
```

16 nanos per call is a high price to pay.   Using the async profiler, we can see stack frames more similar to what you provided:

```
          ns  percent  samples  top
  ----------  -------  -------  ---
   850143389   17.04%       85  java.lang.invoke.MethodHandle.invokeBasic
   579972384   11.62%       58  java.lang.invoke.LambdaForm$MH.0x0000000800c2ec00.invoke
   549984738   11.02%       55  java.lang.invoke.LambdaForm$MH.0x0000000800c2e800.invoke
   469959936    9.42%       47  java.lang.invoke.MethodHandle.asType
   410006230    8.22%       41  io.perfmark.java7.SecretMethodHandleGenerator$MethodHandleGenerator.getGeneration
   369945874    7.41%       37  io.perfmark.java7.jmh_generated.MethodHandleGeneratorBenchmarkTest_MethodHandleGeneratorBenchmark_ifEnabled_jmhTest.ifEnabled_avgt_jmhStub
   319968521    6.41%       32  java.lang.invoke.MutableCallSite.getTarget
   269975606    5.41%       27  java.lang.invoke.Invokers.checkCustomized
   250003085    5.01%       25  io.perfmark.testing.GeneratorBenchmark.getGeneration
   239981399    4.81%       24  io.perfmark.testing.GeneratorBenchmark.isEnabled
   160010885    3.21%       16  java.lang.invoke.Invokers.checkGenericType
   119994550    2.40%       12  java.lang.invoke.MethodHandle.type
   110002564    2.20%       11  io.perfmark.testing.GeneratorBenchmark.ifEnabled
   100003138    2.00%       10  java.lang.invoke.MethodHandle.asTypeCached
   100000295    2.00%       10  java.lang.invoke.LambdaForm$MH.0x0000000800c2e400.invoke_MT
    90054404    1.80%        9  java.lang.invoke.DirectMethodHandle$Holder.invokeVirtual
```

Much of the MethodHandle code has been compiled in or optimized out, but not all of it.  I suspect that you are noticing the overhead because your server code is stuck in C1, waiting for the C2 compiler to free up.  This usually happens while the JIT is optimizing the call paths.   

There is one last, minor possibility: You may be getting stuck in C2 Level 2 or Level 3.    These modes are used to collect profiling information about the functions to be used in the final optimization.  One effect of this is that the methods get *slower* as they get more optimized.   In level 2:

```
Benchmark                                                                           Mode  Cnt   Score   Error  Units
MethodHandleGeneratorBenchmarkTest.MethodHandleGeneratorBenchmark.ifEnabled         avgt    5  18.139 ± 0.337  ns/op
MethodHandleGeneratorBenchmarkTest.MethodHandleGeneratorBenchmark.ifEnabled:·async  avgt          NaN            ---
```

And level 3:
```
    MethodHandleGeneratorBenchmarkTest.MethodHandleGeneratorBenchmark.ifEnabled         avgt    5  63.367 ± 1.737  ns/op
MethodHandleGeneratorBenchmarkTest.MethodHandleGeneratorBenchmark.ifEnabled:·async  avgt          NaN            ---
```

Tier 3 is almost 4x slower than Tier1.   Again, you may be sampling while in this slower mode, which would appear to have a larger cost.



## Alternatives

I strongly recommend that you continue to use the MethodHandleGenerator, as it will generate the lowest overhead steady state.   However, if you suspect that a lot of servers may be stuck in the lower compiler modes for a long time, the other generators offer lower overhead in the lower modes.    I have included the benchmarks for these as a comparison:

JDK6 (VolatileGenerator):
```
Level 4
Benchmark                                                                                Mode  Cnt   Score    Error  Units
VolatileGeneratorBenchmarkTest.VolatileGeneratorBenchmark.getAndSetAndGetGeneration      avgt    5   8.833 ±  1.574  ns/op
VolatileGeneratorBenchmarkTest.VolatileGeneratorBenchmark.getGeneration                  avgt    5   3.010 ±  0.031  ns/op
VolatileGeneratorBenchmarkTest.VolatileGeneratorBenchmark.ifEnabled                      avgt    5   0.961 ±  0.108  ns/op
VolatileGeneratorBenchmarkTest.VolatileGeneratorBenchmark.racyGetAndSetAndGetGeneration  avgt    5  92.225 ± 13.014  ns/op

Level 3
Benchmark                                                                                Mode  Cnt    Score    Error  Units
VolatileGeneratorBenchmarkTest.VolatileGeneratorBenchmark.getAndSetAndGetGeneration      avgt    5   24.459 ±  0.253  ns/op
VolatileGeneratorBenchmarkTest.VolatileGeneratorBenchmark.getGeneration                  avgt    5   13.342 ±  1.695  ns/op
VolatileGeneratorBenchmarkTest.VolatileGeneratorBenchmark.ifEnabled                      avgt    5   12.111 ±  0.120  ns/op
VolatileGeneratorBenchmarkTest.VolatileGeneratorBenchmark.racyGetAndSetAndGetGeneration  avgt    5  444.089 ± 25.959  ns/op

Level 2
Benchmark                                                                                Mode  Cnt    Score   Error  Units
VolatileGeneratorBenchmarkTest.VolatileGeneratorBenchmark.getAndSetAndGetGeneration      avgt    5   18.659 ± 0.182  ns/op
VolatileGeneratorBenchmarkTest.VolatileGeneratorBenchmark.getGeneration                  avgt    5    7.460 ± 0.045  ns/op
VolatileGeneratorBenchmarkTest.VolatileGeneratorBenchmark.ifEnabled                      avgt    5    5.468 ± 0.131  ns/op
VolatileGeneratorBenchmarkTest.VolatileGeneratorBenchmark.racyGetAndSetAndGetGeneration  avgt    5  185.562 ± 7.187  ns/op


Level 1
VolatileGeneratorBenchmarkTest.VolatileGeneratorBenchmark.getAndSetAndGetGeneration      avgt    5   16.315 ± 1.439  ns/op
VolatileGeneratorBenchmarkTest.VolatileGeneratorBenchmark.getGeneration                  avgt    5    5.487 ± 0.351  ns/op
VolatileGeneratorBenchmarkTest.VolatileGeneratorBenchmark.ifEnabled                      avgt    5    4.650 ± 0.048  ns/op
VolatileGeneratorBenchmarkTest.VolatileGeneratorBenchmark.racyGetAndSetAndGetGeneration  avgt    5  111.241 ± 6.842  ns/op

Int
Benchmark                                                                                Mode  Cnt    Score    Error  Units
VolatileGeneratorBenchmarkTest.VolatileGeneratorBenchmark.getAndSetAndGetGeneration      avgt    5  401.126 ± 63.157  ns/op
VolatileGeneratorBenchmarkTest.VolatileGeneratorBenchmark.getGeneration                  avgt    5  256.357 ± 12.439  ns/op
VolatileGeneratorBenchmarkTest.VolatileGeneratorBenchmark.ifEnabled                      avgt    5  247.543 ±  6.721  ns/op
VolatileGeneratorBenchmarkTest.VolatileGeneratorBenchmark.racyGetAndSetAndGetGeneration  avgt    5  488.297 ± 58.877  ns/op
```

JDK9: (VarHandleGenerator):

```
Level 4 
Benchmark                                                                                  Mode  Cnt   Score   Error  Units
VarHandleGeneratorBenchmarkTest.VarHandleGeneratorBenchmark.getAndSetAndGetGeneration      avgt    5   3.732 ± 0.316  ns/op
VarHandleGeneratorBenchmarkTest.VarHandleGeneratorBenchmark.getGeneration                  avgt    5   2.836 ± 0.038  ns/op
VarHandleGeneratorBenchmarkTest.VarHandleGeneratorBenchmark.ifEnabled                      avgt    5   1.133 ± 0.019  ns/op
VarHandleGeneratorBenchmarkTest.VarHandleGeneratorBenchmark.racyGetAndSetAndGetGeneration  avgt    5  22.199 ± 0.833  ns/op

Level 1
Benchmark                                                                                  Mode  Cnt    Score   Error  Units
VarHandleGeneratorBenchmarkTest.VarHandleGeneratorBenchmark.getAndSetAndGetGeneration      avgt    5   19.449 ± 0.197  ns/op
VarHandleGeneratorBenchmarkTest.VarHandleGeneratorBenchmark.getGeneration                  avgt    5    6.763 ± 0.705  ns/op
VarHandleGeneratorBenchmarkTest.VarHandleGeneratorBenchmark.ifEnabled                      avgt    5    5.437 ± 1.135  ns/op
VarHandleGeneratorBenchmarkTest.VarHandleGeneratorBenchmark.racyGetAndSetAndGetGeneration  avgt    5  121.898 ± 4.822  ns/op

Interpreter
Benchmark                                                                                  Mode  Cnt     Score     Error  Units
VarHandleGeneratorBenchmarkTest.VarHandleGeneratorBenchmark.getAndSetAndGetGeneration      avgt    5  2504.110 ±  99.459  ns/op
VarHandleGeneratorBenchmarkTest.VarHandleGeneratorBenchmark.getGeneration                  avgt    5   958.669 ±  10.669  ns/op
VarHandleGeneratorBenchmarkTest.VarHandleGeneratorBenchmark.ifEnabled                      avgt    5   961.614 ±   4.056  ns/op
VarHandleGeneratorBenchmarkTest.VarHandleGeneratorBenchmark.racyGetAndSetAndGetGeneration  avgt    5  2945.757 ± 130.856  ns/op
```

**Comment by ran-su** [comment]
Thank you for this very thorough investigation. 

There is a lot of helpful information in your analysis. I'll verify those in our environment and update here if I have any findings.
